Lib/test/test_htmlparser.py

```diff
@@ -380,6 +380,48 @@ def test_convert_charrefs(self):
         self._run_check('no charrefs here', [('data', 'no charrefs here')],
                         collector=collector())
 
+    def test_convert_charrefs_in_attribute_values(self):
+        # default value for convert_charrefs is now True
+        collector = lambda: EventCollectorCharrefs()
+        self.assertTrue(collector().convert_charrefs)
+
+        # do unescape numeric and hex char refs
+        expected = [('starttag', 'a',
+                     [('href', 'https://example.com?foo¢=bar¢&baz¢=bla¢')]),
+                    ('endtag', 'a')]
+        self._run_check('<a href="https://example.com?foo&#xa2;=bar&#xa2&baz&#162;=bla&#162"></a>', expected, collector=collector())
+
+        # do unescape entity matches not followed by ASCII alphanumeric
+        expected = [('starttag', 'a',
+                     [('href', 'https://example.com?foo¢¢ ¢+¢')]),
+                    ('endtag', 'a')]
+        self._run_check('<a href="https://example.com?foo&cent;&cent &cent+&cent"></a>', expected, collector=collector())
+
+        # do not unescape entity matches followed by ASCII alphanumeric
+        expected = [('starttag', 'a',
+                     [('href', 'https://example.com?foo&center&cent123')]),
+                    ('endtag', 'a')]
+        self._run_check('<a href="https://example.com?foo&center&cent123"></a>', expected, collector=collector())
+
+        # do not unescape entity matches followed by equals
+        expected = [('starttag', 'a',
+                     [('href', 'https://example.com?foo&cent=123')]),
+                    ('endtag', 'a')]
+        self._run_check('<a href="https://example.com?foo&cent=123"></a>', expected, collector=collector())
+
+        # do unescape terminated entity matches followed by equals
+        expected = [('starttag', 'a',
+                     [('href', 'https://example.com?foo¢=123')]),
+                    ('endtag', 'a')]
+        self._run_check('<a href="https://example.com?foo&cent;=123"></a>', expected, collector=collector())
```

Comment:
If possible, it would be better to match the style of the previous test, creating different lists of charrefs (e.g. valid, invalid, named, numeric, etc.) and add them in different places in the attribute (beginning, end, before an alnum/space/semicolon/equal).

Also try to keep the lines shorter than 80 chars (you can remove the initial part of the URLs, since they are not necessary).

Reply:
Thanks, looking at it combining multiple cases in a single attribute is indeed hard to read. I restructured the test to have two scenarios:
- terminated entity, numeric and hex char refs
- unterminated entity char refs

Both include cases for start, middle, end, as well as followed by alphanumeric, non-alphanumeric and equals sign. I hope it's a bit clearer now.

Also updated formatting to respect the 80 char limit.